Create a new blank key sized according to the passphrase

mgfritch opened the pull request:
Padding a passphrase with null chars to arrive at a 32-byte length
later forces a user to also pass null chars via the term when
attempting to manually unlock a subvolume via the fscrypt cli tools.

This also had a side-effect of truncating a longer length passphrase
down to a shorter 32 char length.

fixup for:
https://github.com/SUSE/ceph-csi/commit/cfea8d756231e7eda5150e33a7f5711cfcf3aafb
https://github.com/SUSE/ceph-csi/commit/dd0e1988c08e75ce9564854eab93bb0826a9532d

Signed-off-by: Michael Fritch <mfritch@suse.com>

tserong reviewed (approved):
LGTM. I did wonder about passing a bool or something instead of keySize, to indicate we were in backwards compatibility mode, but I don't know that one is really more readable than the other.

The only thing I'd suggest adding is a comment to the commit to point out that not only does the 32 character limit zero-pad shorter passphrases, but that it also caps the maximum passphrase length at 32 characters.

mgfritch committed:
fscrypt: return `ErrBadAuth` during keyFn retry

fscrypt will infinitely retry the keyFn during an auth failure,
preventing the csi driver from progressing when configured with
an invalid passphrase

See also:
https://github.com/google/fscrypt/blob/8c12cd64ab471d0a73ef4c300d7c40077cad5d5d/actions/callback.go#L102-L106

Signed-off-by: Michael Fritch <mfritch@suse.com>

mgfritch committed:
fscrypt: create a new blank key sized according to the passphrase

Padding a passphrase with null chars to arrive at a 32-byte length
later forces a user to also pass null chars via the term when
attempting to manually unlock a subvolume via the fscrypt cli tools.

This also had a side-effect of truncating any longer length passphrase
down to a shorter 32-byte length.

fixup for:
cfea8d756231e7eda5150e33a7f5711cfcf3aafb
dd0e1988c08e75ce9564854eab93bb0826a9532d

Signed-off-by: Michael Fritch <mfritch@suse.com>
